### PR TITLE
SNOW-823361 Enquote normalized names in constructor of ColumnIdentifier

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/types.py [hunk 1]
```diff
@@ -234,7 +235,7 @@ class ColumnIdentifier:
     """Represents a column identifier."""
 
     def __init__(self, normalized_name: str) -> None:
-        self.normalized_name = normalized_name
+        self.normalized_name = quote_name(normalized_name)
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/types.py:238]: If I'm reading it correctly, this is the core change. Is it correct?

**Review by sfc-gh-jdu — APPROVED** [review]
Thanks!

File: src/snowflake/snowpark/_internal/utils.py [hunk 2]
```diff
@@ -791,3 +791,29 @@ def parse_table_name(table_name: str) -> List[str]:
...
+
+
+EMPTY_STRING = ""
+DOUBLE_QUOTE = '"'
+ALREADY_QUOTED = re.compile('^(".+")$')
+UNQUOTED_CASE_INSENSITIVE = re.compile("^([_A-Za-z]+[_A-Za-z0-9$]*)$")
+
+
+def quote_name(name: str) -> str:
+    if ALREADY_QUOTED.match(name):
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/_internal/utils.py:803]: maybe add docstring to this function? Also, is it a good idea to have this ALREADY_QUOTED check here? It's a bit confusing, because name could be 'A' but maye require additional quoting. Ideally this function should have a 1:1 mapping I feel.

File: src/snowflake/snowpark/_internal/analyzer/expression.py [hunk 3]
```diff
@@ -138,7 +138,7 @@ def with_name(self, new_name: str) -> "Attribute":
             return self
         else:
             return Attribute(
-                analyzer_utils.quote_name(new_name),
+                snowflake.snowpark._internal.utils.quote_name(new_name),
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/expression.py:141]: Should be changed to relative imports.

File: src/snowflake/snowpark/types.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/types.py:238]: Yes, this is the core change. The rest of the changes are (1) moving `quote_name` definition and (2) fixing tests.

**Review by sfc-gh-aalam — COMMENTED** [review]
I would also suggest to run a precommit test with this branch to see if we get any weird edge cases due to different parameter setting in REG environment

**Comment by sfc-gh-stan** [comment]
All tests passed in the precommit: https://buwa.ci1.us-west-2.aws-dev.app.snowflake.com/builds/precommit/ci-dev-133.int.snowflakecomputing.com/13430
